### PR TITLE
PWN-6782 - apply window insets

### PR DESCRIPTION
**Pull request by NikolaiEmelianov**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-6782

## Description of Work

apply window insets, add delegate for control system icons color

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/qr/ui/ScanQrFragment.kt [hunk 1]
```diff
@@ -101,6 +100,10 @@ class ScanQrFragment :
         }
     }
 
+    override fun setSystemBarsStyle() {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/qr/ui/ScanQrFragment.kt:103]: set means that we need to pass some parameter to `set` system bars, some style of other 
i liked the approach with fields more, can't we stick to it?

File: app/src/main/java/org/p2p/wallet/root/DecorSystemBarsDelegate.kt [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    fun setStatusBarStyle(style: SystemIconsStyle) {
+        windowInsetsController.isAppearanceLightStatusBars = when (style) {
+            SystemIconsStyle.BLACK -> true
+            SystemIconsStyle.WHITE -> false
+        }
+    }
+
+    fun setNavigationBarStyle(style: SystemIconsStyle) {
+        windowInsetsController.isAppearanceLightNavigationBars = when (style) {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/root/DecorSystemBarsDelegate.kt:52]: windowInsetsController.isAppearanceLightNavigationBars = style is SystemIconsStyle.BLACK

File: core/src/main/java/org/p2p/core/utils/insets/InsetsExtensions.kt [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+import android.view.inputmethod.InputMethodManager
+
+val systemAndImeType: Int
+    get() = WindowInsetsCompat.Type.systemBars() or WindowInsetsCompat.Type.ime()
+
+fun WindowInsetsCompat.systemAndIme(): Insets = getInsets(systemAndImeType)
+fun WindowInsetsCompat.ime(): Insets = getInsets(WindowInsetsCompat.Type.ime())
+fun WindowInsetsCompat.systemBars(): Insets = getInsets(WindowInsetsCompat.Type.systemBars())
+
+fun View.visibleKeyboard(isVisible: Boolean) {
```
Comment (gslevinkov) [inline core/src/main/java/org/p2p/core/utils/insets/InsetsExtensions.kt:20]: we do have KeyboardExt, is this needed?

File: core/src/main/java/org/p2p/core/utils/insets/InsetsExtensions.kt [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                v.removeOnAttachStateChangeListener(this)
+                v.requestApplyInsets()
+            }
+
+            override fun onViewDetachedFromWindow(v: View) = Unit
+        })
+    }
+}
+
+data class InitialPadding(val left: Int, val top: Int, val right: Int, val bottom: Int)
```
Comment (gslevinkov) [inline core/src/main/java/org/p2p/core/utils/insets/InsetsExtensions.kt:79]: InsetInitialPadding

File: core/src/main/java/org/p2p/core/utils/insets/InsetsExtensions.kt [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+fun recordInitialPaddingForView(view: View) = InitialPadding(
```
Comment (gslevinkov) [inline core/src/main/java/org/p2p/core/utils/insets/InsetsExtensions.kt:81]: explicit return type

**Review by gslevinkov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/common/mvp/BaseFragment.kt [hunk 6]
```diff
@@ -108,31 +114,45 @@ abstract class BaseFragment(@LayoutRes layoutRes: Int) : Fragment(layoutRes), Ba
...
-        with(window) {
-            addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
-            statusBarColor = resourcesProvider.getColor(colorResId)
-            navigationBarColor = resourcesProvider.getColor(navBarColor)
-            if (systemIconsStyle == SystemIconsStyle.WHITE) {
-                decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_VISIBLE
-            } else {
-                decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR
-            }
+    protected open fun applyWindowInsets(rootView: View) {
```
Comment (NikolaiEmelianov) [inline app/src/main/java/org/p2p/wallet/common/mvp/BaseFragment.kt:117]: default implementation

File: app/src/main/java/org/p2p/wallet/common/mvp/BaseFragment.kt [hunk 7]
```diff
@@ -108,31 +114,45 @@ abstract class BaseFragment(@LayoutRes layoutRes: Int) : Fragment(layoutRes), Ba
...
-       Change status bar and its icons color
-       isLight = true - white status bar, dark icons
-       isLight = false - dark status bar, white icons
-       Don't forget to reset status bar color on fragment destroy to restore previous color
-    */
-    protected fun setLightStatusBar(isLight: Boolean) {
-        val window = requireActivity().window
-        val decorView = window.decorView
-        WindowInsetsControllerCompat(window, decorView).isAppearanceLightStatusBars = isLight
+    protected open fun setSystemBarsStyle() {
```
Comment (NikolaiEmelianov) [inline app/src/main/java/org/p2p/wallet/common/mvp/BaseFragment.kt:130]: onResume set default icons colors

File: app/src/main/java/org/p2p/wallet/home/MainFragment.kt [hunk 8]
```diff
@@ -99,11 +103,33 @@ class MainFragment : BaseFragment(R.layout.fragment_main), MainTabsSwitcher, Cen
...
+            )
+            val ime = insets.ime()
+            val bottomNavigationHeight = binding.bottomNavigation.height
+            val bottom = if (ime.bottom > bottomNavigationHeight) bottomNavigationHeight else 0
+            insets.inset(
+                systemBars.left,
+                systemBars.top,
+                systemBars.right,
+                systemBars.bottom + bottom,
+            )
```
Comment (NikolaiEmelianov) [inline app/src/main/java/org/p2p/wallet/home/MainFragment.kt:123]: consume systemBars insets and bottomNavigation.height if ime overlay it

File: app/src/main/java/org/p2p/wallet/root/DecorSystemBarsDelegate.kt [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+import androidx.lifecycle.DefaultLifecycleObserver
+import androidx.lifecycle.LifecycleOwner
+import android.graphics.Color
+import android.os.Build
+
+enum class SystemIconsStyle {
+    BLACK, WHITE
+}
+
+class DecorSystemBarsDelegate constructor(
```
Comment (NikolaiEmelianov) [inline app/src/main/java/org/p2p/wallet/root/DecorSystemBarsDelegate.kt:15]: this class is responsible for system icons colors. For now we have only one theme, style: `SystemIconsStyle.BLACK` and sometimes need to set `SystemIconsStyle.WHITE`
But in future we can add dark theme and default `navigationBarStyle` `statusBarStyle` may change

File: app/src/main/java/org/p2p/wallet/auth/ui/onboarding/root/OnboardingRootFragment.kt [hunk 10]
```diff
@@ -41,6 +43,10 @@ class OnboardingRootFragment :
         popAndReplaceFragment(NewCreatePinFragment.create(), inclusive = true)
     }
 
+    override fun applyWindowInsets(rootView: View) {
+        rootView.doOnApplyWindowInsets { _, insets, _ -> insets }
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/auth/ui/onboarding/root/OnboardingRootFragment.kt:47]: blank function?

File: app/src/main/java/org/p2p/wallet/common/mvp/BaseFragment.kt [hunk 11]
```diff
@@ -71,12 +76,13 @@ abstract class BaseFragment(@LayoutRes layoutRes: Int) : Fragment(layoutRes), Ba
...
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
-        setSystemBarsColors(statusBarColor, navBarColor)
+        applyWindowInsets(view)
     }
 
     override fun onResume() {
         super.onResume()
         logScreenOpenedEvent()
+        setSystemBarsStyle()
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/common/mvp/BaseFragment.kt:85]:
```suggestion
       updateSystemBarsStyle
```

**Review by 1sakov — APPROVED** [review]
Looks good to me. 
Did you test it? 
Did the bottom nav bar aligned to the bottom when keyboard is visible?

**Pull request merged**